src/hotspot/share/gc/shenandoah/heuristics/shenandoahOldHeuristics.cpp

```diff
@@ -0,0 +1,309 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "gc/shenandoah/heuristics/shenandoahOldHeuristics.hpp"
+#include "precompiled.hpp"
+#include "utilities/quickSort.hpp"
+
+ShenandoahOldHeuristics::ShenandoahOldHeuristics(ShenandoahGeneration* generation) :
+    ShenandoahHeuristics(generation),
+    _old_collection_candidates(0),
+    _next_old_collection_candidate(0),
+    _hidden_old_collection_candidates(0),
+    _hidden_next_old_collection_candidate(0),
+    _old_coalesce_and_fill_candidates(0),
+    _first_coalesce_and_fill_candidate(0)
+{
+}
+
+void ShenandoahOldHeuristics::prime_collection_set(ShenandoahCollectionSet* collection_set) {
+  uint included_old_regions = 0;
+  size_t evacuated_old_bytes = 0;
+
+  // TODO: These macro definitions represent a first approximation to desired operating parameters.
+  // Eventually, these values should be determined by heuristics and should adjust dynamically based
+  // on most current execution behavior.  In the interrim, we may choose to offer command-line options
+  // to set the values of these configuration parameters.
+
+  // MAX_OLD_EVACUATION_BYTES represents an "arbitrary" bound on how much evacuation effort is dedicated to
+  // old-gen regions.
+#define MAX_OLD_EVACUATION_BYTES (ShenandoahHeapRegion::region_size_bytes() * 8)
+
+  // PROMOTION_BUDGET_BYTES represents an "arbitrary" bound on how many bytes can be consumed by young-gen
+  // objects promoted into old-gen memory.  We need to avoid a scenario under which promotion of objects
+  // depletes old-gen available memory to the point that there is insufficient memory to hold old-gen objects
+  // that need to be evacuated from within the old-gen collection set.
+  //
+  // TODO We should probably enforce this, but there is no enforcement currently.  Key idea: if there is not
+  // sufficient memory within old-gen to hold an object that wants to be promoted, defer promotion until a
+  // subsequent evacuation pass.  Since enforcement may be expensive, requiring frequent synchronization
+  // between mutator and GC threads, here's an alternative "greedy" mitigation strategy: Set the parameter's
+  // value so overflow is "very rare".  In the case that we experience overflow, evacuate what we can from
+  // within the old collection set, but don't evacuate everything.  At the end of evacuation, any collection
+  // set region that was not fully evacuated cannot be recycled.  It becomes a prime candidate for the next
+  // collection set selection.  Here, we'd rather fall back to this contingent behavior than force a full STW
+  // collection.
+#define PROMOTION_BUDGET_BYTES (ShenandoahHeapRegion::region_size_bytes() / 2)
```

Comment:
I'd put macro definitions at the top of the file, right below the includes.

Reply:
replaced with C++ comment in next commit.